### PR TITLE
Referral engine

### LINKED ISSUE #3645: Referral engine
Add a basic referral engine that generates referral codes for current users to share with new users when signing up to Cal.com

### PR DESCRIPTION
**Pull request by joeauyeung**
## What does this PR do?

This PR adds a referral engine to Cal.com.

Right now this only creates a referral code and tracks referrals.

Linked to https://github.com/calcom/website/pull/185

https://www.loom.com/share/02f15322766348e5833d3e3f29f35510

Fixes # (issue)

**Environment**: Staging(main branch) / Production

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## How should this be tested?

- [ ] Test A
- [ ] Test B

## Checklist

- I haven't read the [contributing guide](https://github.com/calcom/cal.com/blob/main/CONTRIBUTING.md)
- My code doesn't follow the style guidelines of this project
- I haven't performed a self-review of my own code and corrected any misspellings
- I haven't commented my code, particularly in hard-to-understand areas
- I haven't checked if my PR needs changes to the documentation
- I haven't checked if my changes generate no new warnings
- I haven't added tests that prove my fix is effective or that my feature works
- I haven't checked if new and existing unit tests pass locally with my changes

### CODE REVIEW

File: apps/web/pages/refer-a-friend.tsx [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+                  <Button
+                    className="mt-1"
+                    onClick={() => {
+                      sendReferralEmails.mutate({ emails: emailRef.current.value });
+                    }}>
+                    Send
+                  </Button>
+                </div>
+                <p className="mt-2 text-sm text-gray-500" id="email-description">
+                  Separate multiple emails with commas
```
Comment (PeerRich) [inline apps/web/pages/refer-a-friend.tsx:67]: remember to extract strings to `t("string")`

File: apps/web/pages/refer-a-friend.tsx [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+                  <input
+                    value={referralLink}
+                    className="mt-1 mr-2 block w-full rounded-sm border-gray-300 p-1 focus:border-neutral-800 focus:ring-neutral-800 sm:text-sm"
+                    readOnly
+                  />
+                  <Button
+                    className="mt-1"
+                    onClick={() => {
+                      navigator.clipboard.writeText(referralLink);
+                      showToast("Referral link copied");
```
Comment (PeerRich) [inline apps/web/pages/refer-a-friend.tsx:86]: same i18n

File: apps/web/pages/refer-a-friend.tsx [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+                    Copy
+                  </Button>
+                </div>
+              </div>
+            </div>
+
+            <hr className="h-2 border-neutral-200" />
+
+            <h1 className="font-cal mb-1 text-xl font-bold capitalize tracking-wide text-gray-900">
+              Track your referrals
```
Comment (PeerRich) [inline apps/web/pages/refer-a-friend.tsx:97]: same i18n

File: apps/web/pages/refer-a-friend.tsx [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+            </h1>
+            <p className="text-sm text-neutral-500 ltr:mr-4 rtl:ml-4">
+              See how many friends have signed up with your referral
```
Comment (PeerRich) [inline apps/web/pages/refer-a-friend.tsx:100]: same i18n

File: apps/web/pages/refer-a-friend.tsx [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+            </p>
+
+            <hr className="h-2 border-neutral-200" />
+
+            <h3 className="font-cal mb-1 font-bold capitalize tracking-wide text-gray-900">Referral perks</h3>
+            <p className="text-sm text-neutral-500 ltr:mr-4 rtl:ml-4">Check out your referral perks</p>
+          </>
+        ) : (
+          <Loader />
```
Comment (PeerRich) [inline apps/web/pages/refer-a-friend.tsx:109]: maybe a skeleton instead?

**Review by PeerRich — COMMENTED** [review]
left some comments but just saw its still int draft :)

**Comment by Jaibles** [comment]
Designs for this feature are here: https://www.figma.com/file/9MOufQNLtdkpnDucmNX10R/%E2%9D%96-Cal-DS-v2.0?node-id=13363%3A84358

**Comment by joeauyeung** [comment]
@Jaibles or even @shirazdole for the referral email is there a specific design we can do? It feels like it could be a little more "flashier" than our normal booking scheduled emails.

**Comment by Jaibles** [comment]
@joeauyeung we could use this email template I've designed for the drip campaigns?
https://www.figma.com/file/Mt3DdNMRZv4ccuToYuuVWu/Emails?node-id=36%3A187

File: apps/web/pages/refer-a-friend.tsx [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+      if (data) {
+        setReferralLink(
+          `${WEBSITE_URL}/signup/?referralCode=${data?.referrer?.username}${data?.referrer?.referralPin}`
+        );
+
+        setFreeReferees(data?.freeReferees);
+        setProReferees(data?.proReferees);
+      }
+    },
+  });
```
Comment (joeauyeung) [inline apps/web/pages/refer-a-friend.tsx:51]: Something that I can't figure out is sometimes the page loads but the referral link is empty. I know it involves state management but I can't wrap my head around it

File: apps/web/pages/api/referral-code.ts [hunk 7]
```diff
@@ -0,0 +1,29 @@
+import type { NextApiRequest, NextApiResponse } from "next";
+
+import prisma from "@calcom/prisma";
+
+export default async function handler(req: NextApiRequest, res: NextApiResponse): Promise<void> {
+  const { referralCode } = req.body;
+
+  // Minimum length of a referral code is 5 characters
+  if (referralCode.length < 5) return res.status(418).json({ valid: false });
```
[the remark this answers is not in the document]
Comment (zomars) [reply apps/web/pages/api/referral-code.ts:9]: We can use Zod for this

**Comment by zomars** [comment]
I'll copy my comment on Twist here for exposure:

Just a thought. Would it make sense to adopt a “microservice” style for the referral program?

Let me elaborate. 

The referrals program only applies to our hosted Cal app. It doesn’t makes much sense for self hosted users to have this in their codebase does it? 

My proposal is: 
- Have this in the monorepo but…
- Served as another app in `apps/referrals`
- It can be available in `referrals.cal.com` also rewritten in `app.cal.con/referrals` if needed or… be embedded altogether 

Thoughts?

**Comment by zomars** [comment]
Moved to 2.1 to focus on V2

**Comment by PeerRich** [comment]
> 

> Moved to 2.1 to focus on V2

is that a good idea? Thats 30 days from now, likely causing a ton of more merge conflicts. i mean this looks kinda ready, no? @joeauyeung @zomars

**Comment by zomars** [comment]
We don't want this code for self-hosted @PeerRich  we need to change our approach

**Comment by PeerRich** [comment]
> We don't want this code for self-hosted @PeerRich  we need to change our approach 

I wonder if that's actually true. Mabye self hosters want to have a referral engine. like lets say a telehealth marketplace that is looking to attract more therapists

**Comment by zomars** [comment]
But then we would need to maintain and support for it. We don't want to increase core complexity. I stand by my statement.

**Comment by PeerRich** [comment]
> But then we would need to maintain and support for it. We don't want to increase core complexity. I stand by my statement.

sounds good 👌

**Comment by zomars** [comment]
Closing for now

**Pull request closed without merging**